src/nupic/algorithms/SpatialPooler.cpp

```diff
@@ -1120,20 +1122,79 @@ void SpatialPooler::updateDutyCyclesHelper_(vector<Real>& dutyCycles,
 
 void SpatialPooler::updateBoostFactors_()
 {
-  for (UInt i = 0; i < numColumns_; i++)
+  if (maxBoost_ > 1.0)
+  {    
+    if (globalInhibition_)
+    {
+      updateBoostFactorsGlobal_();
+    }
+    else
+    {
+      updateBoostFactorsLocal_();
+    }
+  }
+}
+
+void SpatialPooler::updateBoostFactorsGlobal_()
+{
+  Real targetDensity;
+  if (numActiveColumnsPerInhArea_ > 0)
+  {
+    UInt inhibitionArea = pow((Real) (2 * inhibitionRadius_ + 1),
+                              (Real) columnDimensions_.size());
+    inhibitionArea = min(inhibitionArea, numColumns_);
+    targetDensity = ((Real) numActiveColumnsPerInhArea_) / inhibitionArea;
+    targetDensity = min(targetDensity, (Real) 0.5);
```

Comment:
Here's what I presume. For local inhibition, the `inhibitionRadius_` isn't really a parameter, it changes with the statistics of the data. So if you're using the `numActiveColumnsPerInhArea_` parameter, you can accidentally wind up in situations where you're activating way more columns than you want, because the inhibition areas are small.

This does not apply to global inhibition unless someone goes in and manually changes the inhibition radius. `::initialize` will set this radius to cover the whole space. So in this code, the inhibition radius is predictable. So I don't think we should perform `min` check here.

Similarly, I would argue that this logic in `inhibitColumns_` should be moved to `inhibitColumnsLocal_`, and `inhibitColumnsGlobal_` should just obey the parameters. Though maybe that's out of scope for this change.

Ultimately this is all a hack that exists because the `numActiveColumnsPerInhArea` parameter is awkward when mixed with local inhibition. With local inhibition it's probably best to use the `localAreaDensity` parameter instead.